### PR TITLE
Property address

### PR DESCRIPTION
**Pull request by kru**
#48  
#49
#61

### CODE REVIEW

File: mobile/src/redux/effects/api/http-client.js [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+  sendRequest(requestMethod, url, data = {}) {
+    let headers = data.config ? data.config.headers || {} : {}
+    headers = Object.assign(
+      {},
+      this.client.defaults.headers.common,
+      this.config.headers,
+      headers,
+    )
+
+    return this.client
```
Comment (AaronVasquez) [inline mobile/src/redux/effects/api/http-client.js]: It's probably easier to read if you return a `new Promise` here instead of creating one in the `then` below.

File: mobile/src/redux/effects/api/login.js [hunk 2]
```diff
@@ -0,0 +1,12 @@
+import httpClient from './http-client'
+import { logInActions } from '../../modules/login'
```
Comment (AaronVasquez) [inline mobile/src/redux/effects/api/login.js]: You should never import `actions` into you effects, They should be dispatched in your `epics` only.

File: mobile/src/redux/effects/api/login.js [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+
```
Comment (AaronVasquez) [inline mobile/src/redux/effects/api/login.js]: Create a `logout` function too.

File: mobile/src/redux/effects/api/login.js [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+export const logInRequest = payload => {
+  httpClient.post('/login', payload).then(response => {
+    if (response.data && response.data.access_token) {
+      httpClient.setAuthorizationToken(response.data.access_token)
```
Comment (AaronVasquez) [inline mobile/src/redux/effects/api/login.js]: Lets store the `access_token` on the `user` module and pass it in whenever the API call requires it. We should prefer to use FP techniques instead of storing state in objects :)

1. It makes it more explicit to the caller (the epic) that an API call requires auth.
2. The UI can render different things based on whether an auth_token is present in the `user` module or not.

Comment (AaronVasquez) [inline mobile/src/redux/effects/api/login.js]: `auth_token` should be `authToken`. Since these are communicate using JSON, we should conform to JSON standards.

https://github.com/p4sh4/snaky_camel should make that easy. @EdmundMai is there a "rails way" to do this?

File: mobile/src/redux/modules/login.js [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+    payload,
+  }),
+})
+
+const initState = fromJS({
+  inputField: {
+    email: '',
+    password: '',
+  },
+  isLogin: null,
```
Comment (AaronVasquez) [inline mobile/src/redux/modules/login.js]: We don't need this state. Have your `user` module listen for `LOGIN_ACTIONS.FULFILLED` and set the authToken there.

File: mobile/src/redux/modules/users.js [hunk 6]
```diff
@@ -0,0 +1,26 @@
+import { fromJS } from 'immutable'
+import CONFIG from '../../../config'
+
+export const USERS_ACTIONS = Object.freeze({
+  STORE: `${CONFIG.APP_NAME}/users/store`,
+})
+
+export const userActions = Object.freeze({
+  store: payload => ({
+    type: USERS_ACTIONS.STORE,
```
Comment (AaronVasquez) [inline mobile/src/redux/modules/users.js]: You can share the `LOGIN_ACTIONS.FULFILLED` action instead of creating a separate one.

File: mobile/src/redux/modules/users.js [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+    payload,
+  }),
+})
+
+const initState = fromJS({ userData: null })
```
Comment (AaronVasquez) [inline mobile/src/redux/modules/users.js]: Don't nest this under userData. Just have a user object with `{ email, authToken, etc...}`

File: mobile/src/hoc/to-js.js [hunk 8]
```diff
@@ -0,0 +1,22 @@
+import React from 'react'
+import { Iterable } from 'immutable'
+
+const toJS = WrappedComponent => wrappedComponentProps => {
```
Comment (AaronVasquez) [inline mobile/src/hoc/to-js.js]: Just a note, this hurts performance by having to do a deep conversion everytime the store updates. We have to convert to and from JS to have redux interact with the views.

It might not be worth it to `Immutable` since it adds complexity and hurts performance.

I'll let you decide if you want to keep it or not.

File: mobile/src/redux/effects/api/http-client.js (hunk 1 above)
Comment (AaronVasquez) [reply mobile/src/redux/effects/api/http-client.js]: Looks like you didn't change this?

```js
return new Promise((resolve, reject) => 
      this.client
      .request(
        {
          method: requestMethod,
          url,
          baseURL: this.baseUrl,
          params: data.qs,
          data: data.json || querystring.stringify(data.form) || data.formData,
          headers: Object.assign(
            headers,
            /* eslint-disable */
            data.json
              ? { 'Content-type': 'application/json' }
              : data.form
                ? {
                    'Content-Type': 'application/x-www-form-urlencoded',
                  }
                : { 'Content-Type': 'multipart/form' },
          ),
          /* eslint-enable */
          timeout: 60 * 1000,
          paramsSerializer: params => querystring.stringify(params),
          onUploadProgress: data.config.onUploadProgress || null,
        },
        data.config,
      )
      .then(json => json.message ? reject(json) : resolve(json))
  )
```

File: mobile/src/redux/effects/api/login.js [hunk 9]
```diff
@@ -0,0 +1,25 @@
+import { AsyncStorage } from 'react-native'
```
Comment (AaronVasquez) [inline mobile/src/redux/effects/api/login.js]: @EdmundMai can you show @kru how to use `redux-persist`. We can probably clean this up a little bit.

We can do this in another PR, so this one doesn't get too big.

File: mobile/src/redux/effects/share.android.js [hunk 10]
```diff
@@ -0,0 +1,7 @@
+import { Share } from 'react-native'
+
+export const shareApp = () =>
+  Share.share({
+    message:
+      'Griff has invited you to try a FREE month trial of Blue Collar Lists. To claim your free gift, sign up using this link: https://www.bluecollarlists/shftjvf',
```
Comment (AaronVasquez) [inline mobile/src/redux/effects/share.android.js]: Pass in `firstName` from you user object instead of using `Griff`

File: mobile/src/hoc/to-js.js (hunk 8 above)
Comment (kru) [reply mobile/src/hoc/to-js.js]: This way is the recommendation from redux official documentation. Ok, Let me remove the Immutable first since its have problem with redux-persist too, I've tried integrate the redux persist in this branch before but immutable causing problem.
